### PR TITLE
prevent CSRF attack using Double Submit Cookies solution

### PR DESCRIPTION
**Pull request by zreigz**
**What this PR does / why we need it**:
The implemented solution protects the `kubeconfig` endpoint against CSRF attack using Double Submit Cookies solution.
In this technique, we send a random value in both a cookie and as a request parameter, with the server verifying if the cookie value and request value match.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/handler/kubeconfig.go [hunk 1]
```diff
@@ -10,14 +10,22 @@ import (
...
 	"k8s.io/client-go/tools/clientcmd"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 )
 
+const (
+	csrfCookieName = "csrf_token"
+	cookieMaxAge   = 180
+)
+
+var secureCookie = securecookie.New(securecookie.GenerateRandomKey(32), nil)
```
Comment (mrIncompetent) [inline api/pkg/handler/kubeconfig.go]: We run multiple instances of the API.
Does this lead to decode errors as we have a different random key per instance?

Comment (zreigz) [reply api/pkg/handler/kubeconfig.go]: Good point.
This hash key value should not be also hardcoded, so I will add a new parameter to pass this to API server

File: api/pkg/handler/kubeconfig.go [hunk 2]
```diff
@@ -319,3 +343,20 @@ func (r CreateOIDCKubeconfigReq) GetDC() string {
...
 }
+
+// setCookie add cookie with random string value
+func setCookie(w http.ResponseWriter, nonce string) {
+	value := map[string]string{
+		"nonce": nonce,
+	}
+
+	if encoded, err := secureCookie.Encode(csrfCookieName, value); err == nil {
+		cookie := &http.Cookie{
```
Comment (bashofmann) [inline api/pkg/handler/kubeconfig.go]: For increased security, `SameSite` and `HttpOnly` and `Secure` should be set as well

Comment (zreigz) [reply api/pkg/handler/kubeconfig.go]: Funny thing golang 1.10 doesn't support `SameSite` field. After updating our linter to version 1.11 the CI have passed.

**Comment by zreigz** [comment]
@mrIncompetent @bashofmann PTAL

File: api/pkg/handler/kubeconfig.go [hunk 3]
```diff
@@ -319,3 +358,23 @@ func (r CreateOIDCKubeconfigReq) GetDC() string {
...
 	return r.ProjectID
 }
+
+// setCookie add cookie with random string value
+func setCookie(w http.ResponseWriter, nonce string, secureMode bool, maxAge int) {
+	value := map[string]string{
+		"nonce": nonce,
+	}
+
+	if encoded, err := secureCookie.Encode(csrfCookieName, value); err == nil {
```
Comment (p0lyn0mial) [inline api/pkg/handler/kubeconfig.go]: please handle error

File: api/cmd/kubermatic-api/options.go [hunk 4]
```diff
@@ -70,6 +74,9 @@ func (o serverRunOptions) validate() error {
 		if len(o.oidcIssuerRedirectURI) == 0 {
 			return fmt.Errorf("%s feature is enabled but \"oidc-redirect-uri\" flag was not specified", OIDCKubeCfgEndpoint)
 		}
+		if len(o.oidcIssuerCookieHashKey) == 0 {
```
Comment (p0lyn0mial) [inline api/cmd/kubermatic-api/options.go:77]: should we validate the key ? are there any constraints other than the recommended length (https://github.com/gorilla/securecookie/blob/master/securecookie.go#L124)?

File: api/pkg/handler/kubeconfig.go [hunk 5]
```diff
@@ -187,6 +205,10 @@ type createOIDCKubeconfigRsp struct {
 	phase int
 	// oidcKubeConfig holds not serialized kubeconfig
 	oidcKubeConfig *clientcmdapi.Config
+	// nonce holds an arbitrary number storied in cookie to prevent Cross-site Request Forgery attack.
+	nonce string
+	// cookie received only with HTTPS, never with HTTP.
+	secureCookieMode bool
 }
 
 func encodeKubeconfigDoINeddAcditional(c context.Context, w http.ResponseWriter, response interface{}) (err error) {
```
Comment (p0lyn0mial) [inline api/pkg/handler/kubeconfig.go]: can we change the name ? for example `encodeOIDCKubeconfig` ?

File: api/pkg/handler/kubeconfig.go [hunk 6]
```diff
@@ -284,6 +311,18 @@ func decodeCreateOIDCKubeconfig(c context.Context, r *http.Request) (interface{}
 		if err := json.Unmarshal(rawState, &oidcState); err != nil {
 			return nil, kcerrors.NewBadRequest("incorrect value of state parameter, expected json encoded value, err = %v", err)
 		}
+		// handle cookie when new endpoint is created and secureCookie was initialized
+		if secureCookie != nil {
+			// cookie should be set in initial code phase
+			if cookie, err := r.Cookie(csrfCookieName); err == nil {
+				value := make(map[string]string)
```
Comment (p0lyn0mial) [inline api/pkg/handler/kubeconfig.go]: does it have to be a `map` ?  how about passing just a `string` ?

**Review by p0lyn0mial — CHANGES_REQUESTED** [review]
nice pull, I have a few minor comments and this lgtm.

File: api/pkg/handler/kubeconfig.go (hunk 6 above)
Comment (zreigz) [reply api/pkg/handler/kubeconfig.go]: done

File: api/pkg/handler/kubeconfig.go (hunk 5 above)
Comment (zreigz) [reply api/pkg/handler/kubeconfig.go]: done

File: api/cmd/kubermatic-api/options.go (hunk 4 above)
Comment (zreigz) [reply api/cmd/kubermatic-api/options.go:77]: It is up to the operator. We don't have any restriction in the code for secrets. It is rather a recommendation. I added an additional description for options parameter

File: api/pkg/handler/kubeconfig.go (hunk 3 above)
Comment (zreigz) [reply api/pkg/handler/kubeconfig.go]: done

**Comment by zreigz** [comment]
@p0lyn0mial thanks for the review, PTAL

**Review by p0lyn0mial — APPROVED** [review]

**Pull request merged**